Clear frame-specific storage access state when WebKitTestRunner revokes permission

charliewolfe opened the pull request:
#### d0700078af9d0e4945391a283eaad15fd45273a5
<pre>
Clear frame-specific storage access state when WebKitTestRunner revokes permission
<a href="https://bugs.webkit.org/show_bug.cgi?id=297491">https://bugs.webkit.org/show_bug.cgi?id=297491</a>
<a href="https://rdar.apple.com/158441742">rdar://158441742</a>

Reviewed by Rupin Mittal.

In 298743@main, I changed the Storage Access API to grant storage access at the frame level instead of
the page level. This undid the changes introduced in 298535@main, which cleared page-level permission
state when WKTR revoked permission. We should also be clearing frame-specific storage state in this case.

This change progresses the following WPTs, though we won’t have CI coverage because we lack a local
DNS resolver:
imported/w3c/web-platform-tests/storage-access-api/requestStorageAccess-cross-site-iframe.sub.https.window.html
imported/w3c/web-platform-tests/storage-access-api/requestStorageAccess-nested-cross-site-iframe.sub.https.window.html

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::revokeFrameSpecificStorageAccess):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::revokeFrameSpecificStorageAccess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::clearResourceLoadStatistics):

Canonical link: <a href="https://commits.webkit.org/298792@main">https://commits.webkit.org/298792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ab0ec467b11faf5dfa34db89717dd70884621a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116679 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26905 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122753 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/67251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f2b70291-ab46-4145-9526-7f70e4886d0d) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118568 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37041 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44932 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/67251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2b373fe4-4357-4d4d-bea9-33646760120b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119628 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29537 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104664 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69080 "Passed tests") | | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/28599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22770 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66420 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98917 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22926 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125889 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43578 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43942 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100866 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/97079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42396 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/20321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18630 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43464 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49059 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42931 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46270 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/44636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebCore/loader/LocalFrameLoaderClient.h:
```diff
@@ -342,6 +342,7 @@ class WEBCORE_EXPORT LocalFrameLoaderClient : public FrameLoaderClient {
 #endif
 
     virtual bool hasFrameSpecificStorageAccess() { return false; }
+    virtual void revokeFrameSpecificStorageAccess() { }
```
Could make this pure virtual.

charliewolfe replied:
I could, but then I'd need to add a blank function to the WebFrameLoaderClient in WebKitLegacy, and this isn't supported there.

RupinMittal replied:
Ah true, yeah it's fine as-is then.

RupinMittal reviewed (approved):
If storage access is now per-frame, should we remove the page level access (e.g. `clearPageLevelStorageAccess`)?

charliewolfe committed:
Clear frame-specific storage access state when WebKitTestRunner revokes permission
https://bugs.webkit.org/show_bug.cgi?id=297491
rdar://158441742

Reviewed by Rupin Mittal.

In 298743@main, I changed the Storage Access API to grant storage access at the frame level instead of
the page level. This undid the changes introduced in 298535@main, which cleared page-level permission
state when WKTR revoked permission. We should also be clearing frame-specific storage state in this case.

This change progresses the following WPTs, though we won’t have CI coverage because we lack a local
DNS resolver:
imported/w3c/web-platform-tests/storage-access-api/requestStorageAccess-cross-site-iframe.sub.https.window.html
imported/w3c/web-platform-tests/storage-access-api/requestStorageAccess-nested-cross-site-iframe.sub.https.window.html

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::revokeFrameSpecificStorageAccess):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::revokeFrameSpecificStorageAccess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::clearResourceLoadStatistics):

Canonical link: https://commits.webkit.org/298792@main
